### PR TITLE
[kotlin2cpg] Improve fullnames.

### PR DESCRIPTION
**Pull request by ml86**
Improve and cleanup the fullname calculation for methods and callsites.
The most noteable changes here are:
1. The lambda function naming schema changed from
   `package.<fileName>.<lambda>counter:signature` to
   `package.scope.<lambda>counter:signature`.
   This give a better indication which lambda function is represented
   also gets rid of exception of using the filename in the method
   fullnames.
2. Fixed lambda function signatures.
3. The fullname schema of extension methods changed from
   `extendedClassFullName.extensionMethodName:signatureWithoutThisParameter`
   to
   `package.extensionMethodName:signatureWithThisParameter`
   The fact that the this parameter is part of the signature is an
   exception in our java fullnames but necessary in order to distinguish
   otherwise identical extension methods.
4. We avoid the recalculation of the binding context with every file
   processed. This is now done once at the begining.
5. Calls to `::class` are now represented as calls to
   `<operator>.class`.
6. Calculate proper BINDING nodes for lambda and anonymous functions.
7. Calculate proper base class for type decls created for lambda and anonymous functions.
8. Fully removed old TypeRenderer class. All full names now go via new rendering.
9. `--keep-type-arguments` has no effect anymore. This feature was only implemented partially and
    we might bring it back later.

### CODE REVIEW

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala [hunk 1]
```diff
@@ -204,12 +204,13 @@ trait AstForPrimitivesCreator(implicit withSchemaValidation: ValidationMode) {
     argName: Option[String],
     annotations: Seq[KtAnnotationEntry] = Seq()
   )(implicit typeInfoProvider: TypeInfoProvider): Ast = {
-    val (fullName, signature) = typeInfoProvider.fullNameWithSignature(expr, ("", "")) // TODO: fix the fallback names
-    val typeFullName          = registerType(typeInfoProvider.expressionType(expr, TypeConstants.javaLangObject))
+    val typeFullName = registerType(typeInfoProvider.expressionType(expr, TypeConstants.javaLangObject))
+    val fullName     = "<operator>.class"
```
Comment (maltek) [inline joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala:208]: wouldn't a `TypeRef` be a great fit for `::class`?

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/types/NameRenderer.scala [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+          }
+      }
+
+    kotlinFullName.map { kotlinFullName =>
+      val javaFullName = builtinTypeTranslationTable.get(kotlinFullName)
+      if (javaFullName.isDefined) {
+        javaFullName.get
+      } else if (kotlinFullName.startsWith("kotlin.Function")) {
+        // For historical reasons kotlin.Function types are not mapped via JavaToKotlinClassMap.
+        // In order to not touch the policies, we mimic that for now.
```
Comment (maltek) [inline joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/types/NameRenderer.scala:142]: I think we should make the switch. We don't have too many such policies yet, and we can write a simple script to duplicate them during the migration.

File: joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala [hunk 3]
```diff
@@ -11,25 +11,34 @@ class ExtensionTests extends KotlinCode2CpgFixture(withOssDataflow = false) {
...
         |""".stripMargin)
 
     "should contain a CALL node for the calls to the extension fns with the correct MFN set" in {
       val List(c) = cpg.call.code(".*printBaz.*").l
-      c.methodFullName shouldBe "mypkg.Example.printBaz:void()"
+      c.methodFullName shouldBe "mypkg.printBaz:void(mypkg.Example,java.lang.String)"
     }
 
     "should contain a METHOD node for the extension fn with the correct MFN set" in {
+      val x       = cpg.method.fullName.l
```
Comment (maltek) [inline joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala:29]:
```suggestion
```

File: joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala [hunk 4]
```diff
@@ -11,25 +11,34 @@ class ExtensionTests extends KotlinCode2CpgFixture(withOssDataflow = false) {
...
       val List(m) = cpg.method.fullName(".*printBaz.*").l
-      m.fullName shouldBe "mypkg.Example.printBaz:void()"
+      m.fullName shouldBe "mypkg.printBaz:void(mypkg.Example,java.lang.String)"
+    }
+
+    "should contain a METHOD node for the extension fn with the correct parameter indicies" in {
+      val x = cpg.method.fullName.l
```
Comment (maltek) [inline joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala:35]:
```suggestion
```

**Review by maltek — COMMENTED** [review]
looks much better than before :)

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala [hunk 5]
```diff
@@ -78,7 +78,17 @@ trait AstForExpressionsCreator(implicit withSchemaValidation: ValidationMode) {
     val (fullName, signature) =
       if (operatorOption.isDefined) (operatorOption.get, TypeConstants.any)
       // TODO: fix the fallback METHOD_FULL_NAME and SIGNATURE here (should be a correct number of ANYs)
```
[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala:80]: It actually does not there is just `TypeConstants.any` used as a fallback if fullname and/or signature cannot be obtained via the descriptors.

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala [hunk 6]
```diff
@@ -454,9 +485,22 @@ trait AstForExpressionsCreator(implicit withSchemaValidation: ValidationMode) {
     val explicitSignature = s"${TypeConstants.any}(${argAsts.map { _ => TypeConstants.any }.mkString(",")})"
     val explicitFullName =
       if (typeInfoProvider.typeRenderer.keepTypeArguments && typeArgs.nonEmpty)
-        s"$methodFqName<${typeArgs.mkString(",")}>:$explicitSignature"
-      else s"$methodFqName:$explicitSignature"
-    val (fullName, signature) = typeInfoProvider.fullNameWithSignature(expr, (explicitFullName, explicitSignature))
+        s"$methodFqName<${typeArgs.mkString(",")}>"
```
[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala:488]: I just did not want to touch this for now. There are lots and lots of places in kotlin2cpg that need fixing but I need to stop somewhere for now.

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala (hunk 1 above)
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala:208]: No, `::class` can give you the class object for some like a class name `MyClass::class` but also for any object like `someObject::class`. In the later case there is no static information about a type to populate a TypeRef. A TypeRef is used for the `MyClass` identifier though.

File: joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala (hunk 3 above)
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/test/scala/io/joern/kotlin2cpg/querying/ExtensionTests.scala:29]: Done

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/Kotlin2Cpg.scala [hunk 7]
```diff
@@ -315,4 +321,19 @@ class Kotlin2Cpg extends X2CpgFrontend[Config] with UsesService {
...
     } yield FileContentAtPath(fileContents, relPath, fileName)
   }
+
+  private def createBindingContext(environment: KotlinCoreEnvironment): BindingContext = {
+    try {
+      logger.info("Running Kotlin compiler analysis...")
+      val t0             = System.currentTimeMillis()
+      val analysisResult = KotlinToJVMBytecodeCompiler.INSTANCE.analyze(environment)
+      val t1             = System.currentTimeMillis()
+      logger.info(s"Kotlin compiler analysis finished in `${t1 - t0}` ms.")
```
[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/Kotlin2Cpg.scala:331]: Done

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/types/NameRenderer.scala (hunk 2 above)
Comment (ml86) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/types/NameRenderer.scala:142]: Done.

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (maltek) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForExpressionsCreator.scala:488]: OK, makes sense.

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala (hunk 1 above)
Comment (maltek) [reply joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/AstForPrimitivesCreator.scala:208]: I see, makes sense now.

**Review by maltek — APPROVED** [review]

**Review by maltek — APPROVED** [review]

**Pull request merged**